docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,167 @@
+
+Another benefit provided by initializers is being able to compact verbose guard clauses into a singular if statement.   
+
+Example:
+
+```lua
+function PlayerControl:GetEntities()
+    if self.RemovedSelf then
+        return self.Instances
+    end
+
+    local entity = self:TryGetEntity()
+    if entity == nil then
+        return self.Instances
+    end
+
+    local index = table.find(self.Instances, entity.Root)
+    if index == nil then
+        return self.Instances
+    end
+
+    table.remove(self.Instances, index)
+    self.RemovedSelf = true
+
+    return self.Instances
+end
+```
+
+```lua
+function PlayerControl:GetEntities()
+    if self.RemovedSelf then
+    elseif local entity = self:TryGetEntity() in entity == nil then
+    elseif local index = table.find(self.Instances, entity.Root) then
+        table.remove(self.Instances, index)
+        self.RemovedSelf = true
+    end
+
+    return self.Instances
+end
+```
+
+# Design
+
+If statements with initializers must match the below grammar. The variables declared by an initializer are only available to the if statement's blocks; any code after the if statement won't have the variables defined.
+
+```diff
+  stat = varlist '=' explist |
+      ...
+      'repeat' block 'until' exp |
+-     'if' exp 'then' block {'elseif' exp 'then' block} ['else' block] 'end' |
++     'if' cond 'then' block {'elseif' cond 'then' block} ['else' block] 'end' |
+      'for' binding '=' exp ',' exp [',' exp] 'do' block 'end' |
+      ...
++ cond = 'local' binding '=' exp ['in' exp] |
++     'local' bindinglist '=' explist 'in' exp |
++     exp
+```
+
+In the former case, the value of the first declared variable will be checked.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() then
+   print(b, "truthy block")
+else
+   print(b, "falsy block")
+end
+```
+
+`Output: true	truthy block`
+
+In the latter case, the `exp` condition is checked rather than the initializer.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() in b == false then
+   print(b, "truthy block")
+else
+   print(b, "falsy block")
+end
+```
+
+`Output: true	falsy block`
+
+When declaring multiple values inside of an initializer, the `in` clause is required.
+
+Example:
+
+```lua
+local function foo()
+    return true, false
+end
+
+if local a, b = foo() in a and b then
+else
+    print'Hello World, from Luau!'
+end
+```
+
+`Output: Hello World, from Luau!`
```

Comment:
With an eye towards consistent behavior with destructuring in the future, I would prefer to see this changed to requiring `a` and `b` be non-nil implicitly, but it's ultimately okay if we air on the conservative side here since this requirement is forwards-compatible with changing it in the future to mean that. In general, I think the behavior we want though is that using `local a = foo()` and `local a, b = foo()` in a condition will only ever take that branch when the results are non-nil.

Reply:
That makes sense. If you want to allow `b` to either exist or not, you should probably have to say that explicitly. Otherwise the fact that you asked for `local a, b = foo()` probably means that you wanted both `a` and `b` to both exist.